File: garch/garch.py
"""GARCH LLH benchmark, Python version."""
import os

import math
import time
import numpy as np
import json
import argparse
from numba import jit
from util import benchmark, garch_data

parser = argparse.ArgumentParser()
parser.add_argument('-n', metavar='n', default=1000, type=int,
                    help='number of iterations')
parser.add_argument('-mode', metavar='mode', default='std',
                    choices=['std', 'numba', 'c++'], help='use the Numba jit')
args = parser.parse_args()


def garchSim(ret2, p):
    h = np.zeros_like(ret2)
    h[0] = np.mean(ret2)
    for i in range(1, len(ret)):
        h[i] = p[0] + p[1]*ret2[i-1] + p[2]*h[i-1]
    return h


def garchLLH(y, par):
    h = garchSim(np.square(y), par)
    t = len(y)
    return -0.5*(t-1)*np.log(2*math.pi)-0.5*np.sum(np.log(h)+(y/np.sqrt(h))**2)


if args.mode == 'numba':
    garchSim = jit(garchSim)
    garchLLH = jit(garchLLH)

out = {}
ret, x0, val_llh = garch_data()


if args.mode == 'c++':
    import cppimport
    cpp_llh = cppimport.imp('cppllh')
    out['c++'] = benchmark(lambda: cpp_llh.garchLLH(ret, x0), args.n, val_llh)
    print('c++ time:', out['c++'])
else:
    out['numpy-' + args.mode] = benchmark(lambda: garchLLH(ret, x0), args.n,
                                          val_llh)
print(json.dumps(out))

File: garch/garch.jl
"GARCH LLH benchmark, Julia version."

import JSON
using Test
using Statistics

data = JSON.parsefile("data.json")
out = Dict()

function garchSim(ɛ²::Vector, ω, α, β)
    h = similar(ɛ²)
    h[1] = mean(ɛ²)
    for i = 2:length(ɛ²)
        @inbounds h[i] = ω + α*ɛ²[i-1] + β*h[i-1]
    end
    h
end

function garchLLH(y::Vector, params::Vector)
    ɛ² = y.^2
    T = length(y)
    h = garchSim(ɛ², params...)
    -(T-1)*log(2π)/2 - sum(log.(h) + (y./sqrt.(h)).^2)/2
end

function benchmark(n)
    #price = Array{Float32}(data["price"])
    ret = Array{Float32}(data["ret"])
    #ret = diff(log.(price))
    #ret = ret .- mean(ret)
    x0 = Array{Float32}(data["x0"])
    tmp = garchLLH(ret, x0)  # warm-up
    @test tmp ≈ data["llh"] atol=1e-3
    t = @elapsed for i = 1:n; tmp = garchLLH(ret, x0); end
end

n_ix = findfirst(ARGS.=="-n")
if n_ix < length(ARGS)
    n = tryparse(Int64, ARGS[n_ix+1])
    n==nothing && error("Wrong number of iterations!")
else
    error("Number of iterations is not set!")
end

out["julia"] = benchmark(n)
println(JSON.json(out))